materials_io/base.py

```diff
@@ -59,41 +58,57 @@ def parse(self, group: List[str], context: dict = None) -> dict:
             (dict): The parsed results, in JSON-serializable format.
         """
 
-    def parse_as_unit(self, files: List[str]) -> dict:
+    def parse_as_unit(self, files, context: dict = None) -> dict:
         """Parse a group of files and merge their metadata
 
         Used if each file in a group are parsed separately, but the resultant metadata
         should be combined after parsing.
 
         Args:
-            files ([str]): List of files to parse
+            files (str or [str]): Path(s) to parse
+            context (dict): Context about the files
         Returns:
-            (dict): Metadata summary from
+            (dict): Metadata summary from the files
         """
-
         # Initialize output dictionary
         metadata = {}
 
         # Loop over all files
-        for group in self.group(files):
+        for group in self.group(files, context):
             try:
-                record = self.parse(group)
+                record = self.parse(group, context)
             except Exception:
                 continue
-            metadata = dict_merge(metadata, record)
+            else:
+                metadata = dict_merge(metadata, record, append_lists=True)
         return metadata
 
-    def group(self, files: Iterable[str], context: dict = None) -> Iterator[Tuple[str, ...]]:
+    def group(self, files, context: dict = None) -> Iterator[Tuple[str, ...]]:
         """Identify a groups of files that should be parsed together
 
         Args:
-            files ([str]): List of paths, which could include both files and directories
+            files (str or [str]): Path(s), which could include both files and directories
             context (dict): Context about the files
         Yields:
             ((str)): Groups of files
         """
-
-        return zip(f for f in files if not os.path.isdir(f))
+        if isinstance(files, str):
+            files = [files]
+        # Clean paths
+        files = [os.path.abspath(os.path.expanduser(f)) for f in files]
+
+        # Default: Every file is in its own group, skipping symlinks
+        groups = []
+        for fn in files:
+            if os.path.isfile(fn):
+                groups.append((fn,))
+            # Recurse through directories (could also be done with os.walk(), especially
+            # for groups that may span directories)
+            elif os.path.isdir(fn):
+                for f in os.listdir(fn):
+                    groups.extend(self.group(os.path.join(fn, f), context))
+
+        return groups
```

Comment:
How about switching this to a generator rather than returning a large list?

I'm concerned about what happens with this code on a huge directory.

Reply:
Sure, that's reasonable.